### PR TITLE
Improve case operation serialization

### LINKED ISSUE #1286: Unable to use custom UserType in when
I'm not able to use custom UserType in when statement in **3.6.0+**. The following exception is always thrown.

```
Exception in thread "main" java.lang.IllegalArgumentException: Unsupported constant Approved
```

It works fine with **3.5.1**.

sample query:
```java
QRequest r = QRequest.request;
JPAQuery q = new JPAQuery(em);
q.from(r);
q.orderBy(r.status.when(new Approved()).then(0).otherwise(1).asc());
q.list(r);
```
status field is defined as:
```java
...
@Type(type = "com.xyz.status.StatusType")
private RequestStatus status;
...
```

StatusType is custom UserType which stores java object to database as string 

```java
...
public void nullSafeSet(PreparedStatement st, Object value, int index, SessionImplementor session) throws HibernateException, SQLException {
		if (value == null) {
			st.setNull(index, StringType.INSTANCE.sqlType());
		} else {
			st.setString(index, value.getClass().getSimpleName());
		}
	}
...
```

### PR DESCRIPTION
**Pull request by timowest**
Fixes #1286
backport https://github.com/querydsl/querydsl/pull/1299

### CODE REVIEW

File: querydsl-jpa/src/main/java/com/querydsl/jpa/JPQLSerializer.java [hunk 1]
```diff
@@ -337,7 +340,7 @@ public void visitLiteral(Object constant) {
         } else if (constant instanceof Enum) {
```
Comment (Shredder121) [inline querydsl-jpa/src/main/java/com/querydsl/jpa/JPQLSerializer.java]: For consistency the `instanceof` can also be `Enum<?>`.

File: querydsl-jpa/src/test/java/com/querydsl/jpa/AbstractJPATest.java [hunk 2]
```diff
@@ -319,6 +319,11 @@ public void Case3() {
     }
 
     @Test
+    public void Case4() {
+        query().from(cat).list(cat.mate.when(savedCats.get(0)).then(0).otherwise(1));
```
Comment (johnktims) [inline querydsl-jpa/src/test/java/com/querydsl/jpa/AbstractJPATest.java]: You might want to assert something here.

Also, you added a test named `Case4` to #1305 .

**Comment by timowest** [comment]
I added some assertions.

**Comment by Shredder121** [comment]
EclipseLink seems to return a `Vector` of `Long`s.

**Comment by timowest** [comment]
Yes, I am working on it right now.

**Comment by timowest** [comment]
I put the fix here https://github.com/querydsl/querydsl/pull/1305

**Comment by Shredder121** [comment]
@johnktims This one looks good to me.

**Pull request merged**